develop/data-generation/block-models.md

```diff
@@ -0,0 +1,315 @@
+
+::: info PREREQUISITES
+You'll have to register the blocks, to which you want the data to be generated. Also see the [blockstate](../blocks/blockstates.md) and [first block](../blocks/first-block.md) tutorials.
+:::
+
+```java
+@Override
+public void generateBlockStateModels(BlockStateModelGenerator blockStateModelGenerator) {
+}
+```
+
+This part of the tutorial will happen entirely in the `generateBlockStateModels` method. Notice the parameter `BlockStateModelGenerator blockStateModelGenerator` - this class will be responsible for generating all the JSON files.
+Here are some handy examples you can use to generate your desired models:
+
+### Simple Cube All {#simple-cube-all}
+
+@[code lang=java transcludeWith=:::datagen-model:cube-all](@/reference/latest/src/client/java/com/example/docs/datagen/FabricDocsReferenceModelProvider.java)
+
+The easiest and most basic method there is. It generates a JSON model file for a normal cubic block. It provides the same texture (named `steel_block`) for all six sides.
+
+```json
+{
+  "parent": "minecraft:block/cube_all",
+  "textures": {
+    "all": "fabric-docs-reference:block/steel_block"
+  }
+}
+```
+
+It also generates a blockstate JSON file. Since we have no blockstates (e.g. Axis, Facing, ...), one variant is sufficient, and is used everytime the block is placed.
+
+```json
+{
+  "variants": {
+    "": {
+      "model": "fabric-docs-reference:block/steel_block"
+    }
+  }
+}
+```
+
+### Singletons {#singletons}
+
+The `registerSingleton` method provides JSON model files based on the `TexturedModel` you pass in and a single blockstate variant. For example
+
+@[code lang=java transcludeWith=:::datagen-model:cube-top-for-ends](@/reference/latest/src/client/java/com/example/docs/datagen/FabricDocsReferenceModelProvider.java)
+
+will generate models for a normal cube, that uses a PNG file `pipe_block` for the sides and a PNG file `pipe_block_top` for the top and bottom sides.
+
+```json
+{
+  "parent": "minecraft:block/cube_column",
+  "textures": {
+    "end": "fabric-docs-reference:block/pipe_block_top",
+    "side": "fabric-docs-reference:block/pipe_block"
+  }
+}
+```
+
+:::tip
+If you're stuck choosing which TextureModel you should use, open the `TexturedModel` class and look at the `TextureMaps`!
+:::
+
+### Block Texture Pool {#block-texture-pool}
+
+@[code lang=java transcludeWith=:::datagen-model:block-texture-pool-normal](@/reference/latest/src/client/java/com/example/docs/datagen/FabricDocsReferenceModelProvider.java)
+
+Another useful method is `registerCubeAllModelTexturePool`. Define the textures by passing in the "base block", and then append the "children", which will have the same textures. <br>
+In this case, we passed in the `RUBY_BLOCK`, so the stairs, slab and fence will use the `RUBY_BLOCK` texture.
+***It will also generate a [simple cube all JSON model](#simple-cube-all) for the "base block" to ensure the texture provision!*** <br>
```

Comment:
could you please clarify/rewrite the "...to ensure texture provision"? don't understand what you meant here.